### PR TITLE
Added support for GPTJ with Lora linear layers

### PR DESCRIPTION
**Pull request by ChiragBSavani**
Support for lora adapter with GPT-J.
Changes are similar to GPT-2 architecture.

### CODE REVIEW

File: src/transformers/adapters/models/gptj/adapter_model.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+
+        Args:
+            head_name (str): The name of the head.
+            overwrite_ok (bool, optional): Force overwrite if a head with the same name exists. Defaults to False.
+        """
+        head = CausalLMHead(self, head_name)
+        self.add_prediction_head(head, overwrite_ok=overwrite_ok)
+
+
+class GPTJModelWithHeads(GPTJAdapterModel):
```
Comment (calpt) [inline src/transformers/adapters/models/gptj/adapter_model.py:186]: The model class of the form XModelWithHeads are deprecated, so we don't intend to add these classes for newly supported architectures. Instead the `GPTJAdapterModel` class should always be used.

File: src/transformers/models/gptj/modeling_gptj.py [hunk 2]
```diff
@@ -116,8 +119,8 @@ def __init__(self, config):
         self.scale_attn = torch.sqrt(torch.tensor(self.head_dim, dtype=torch.float32)).to(torch.get_default_dtype())
 
         self.k_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=False)
-        self.v_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=False)
-        self.q_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=False)
+        self.v_proj = LoRALinear(self.embed_dim, self.embed_dim, "selfattn", config, bias=False)
+        self.q_proj = LoRALinear(self.embed_dim, self.embed_dim, "selfattn", config, bias=False)
```
Comment (calpt) [inline src/transformers/models/gptj/modeling_gptj.py:123]: The module instantiation of the `LoRALinear` layers have changed slightly very recently. See e.g. https://github.com/adapter-hub/adapter-transformers/blob/9eb70923ce5f40169b3176b8666d1424ded8ab3a/src/transformers/models/bert/modeling_bert.py#L270-L272


```suggestion
        self.v_proj = LoRALinear(self.embed_dim, self.embed_dim, "selfattn", config, attn_key="v", bias=False)
        self.q_proj = LoRALinear(self.embed_dim, self.embed_dim, "selfattn", config, attn_key="q", bias=False)
```

File: src/transformers/models/gptj/modeling_gptj.py [hunk 3]
```diff
@@ -116,8 +119,8 @@ def __init__(self, config):
         self.scale_attn = torch.sqrt(torch.tensor(self.head_dim, dtype=torch.float32)).to(torch.get_default_dtype())
 
         self.k_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=False)
```
Comment (calpt) [inline src/transformers/models/gptj/modeling_gptj.py:121]:
```suggestion
        self.k_proj = LoRALinear(self.embed_dim, self.embed_dim, "selfattn", attn_key="k", bias=False)
```

**Review by calpt — COMMENTED** [review]
Hey @ChiragBSavani, thank you very much for your contribution. Looks good so far, I've added a few comments.
To have this merged, it would be useful to add testing and documentation as described in our contribution guide at https://docs.adapterhub.ml/contributing/adding_adapters_to_a_model.html. It would be great if you could have a look at that. Let me know if anything is unclear from that document or in general!

File: src/transformers/adapters/models/gptj/adapter_model.py (hunk 1 above)
Comment (ChiragBSavani) [reply src/transformers/adapters/models/gptj/adapter_model.py:186]: Alright let me remove this class definition then.

**Pull request closed without merging**